File: C++_Version/121Best_Time_to_Buy_and_Sell_Stock.cpp
//Say you have an array for which the ith element is the price of a given stock on day i.
//
//If you were only permitted to complete at most one transaction(ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
//
//Example 1:
//Input: [7, 1, 5, 3, 6, 4]
//	Output : 5
//
//	max.difference = 6 - 1 = 5 (not 7 - 1 = 6, as selling price needs to be larger than buying price)
//	Example 2 :
//	Input : [7, 6, 4, 3, 1]
//	Output : 0
//
//	In this case, no transaction is done, i.e.max profit = 0.

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
using namespace std;

class Solution {
public:
	int maxProfit(vector<int>& prices) {
		if (prices.size() <= 1)
			return 0;
		
		int min = prices[0];
		int size = prices.size();
		int delta = 0,delta_max = 0;

		for (int i = 1; i < size; ++i)
		{
			if (prices[i] <= min)
				min = prices[i];

			else
			{
				delta = prices[i] - min;
				if (delta > delta_max)
					delta_max = delta;
			}
		}

		return delta_max;
	}
};


int main()
{
	vector<int> vec1({ 7, 6, 4, 3, 1 });

	Solution sol1;
	int ret1 = sol1.maxProfit(vec1);

	/*for (int i = 0; i < (int)number.size(); ++i)
		cout << number[i] << ' ';*/

	system("pause");

	return 0;
}

File: Java_Version/121Best_Time_to_Buy_and_Sell_Stock.java
/*Say you have an array for which the ith element is the price of a given stock on day i.

        If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.

        Example 1:
        Input: [7, 1, 5, 3, 6, 4]
        Output: 5

        max. difference = 6-1 = 5 (not 7-1 = 6, as selling price needs to be larger than buying price)
        Example 2:
        Input: [7, 6, 4, 3, 1]
        Output: 0

        In this case, no transaction is done, i.e. max profit = 0.*/

import java.util.*;
import java.lang.Math;
import java.lang.System;
import java.lang.Integer;


public class Main {

    public static void main(String[] args) throws ArithmeticException {

        int[] input = {7,1,5,3,6,4};

        Solution solution = new Solution();

        int result = solution.maxProfit(input);

        System.out.println("haha");
    }

}


class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}


class Solution {
    public int maxProfit(int[] prices) {

        if (prices == null || prices.length < 2) {
            return 0;
        }

        int pricesLength = prices.length;

        int result = 0;
        int minPrice = prices[0];

        for (int i = 1; i < pricesLength; i++) {

            if (prices[i] > minPrice) {
                result = Math.max(result, prices[i] - minPrice);
            } else {
                minPrice = prices[i];
            }

        }

        return result;
    }
}